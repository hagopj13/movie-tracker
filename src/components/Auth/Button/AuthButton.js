// @flow
import React from 'react';
import { connect } from 'react-redux';
import { createStructuredSelector } from 'reselect';
import { makeStyles } from '@material-ui/core/styles';
import Button from '@material-ui/core/Button';

import DialogTypes from 'components/Dialog/types';
import authSelectors from 'store/auth/auth.selectors';
import authActions from 'store/auth/auth.actions';
import AuthActionTypes from 'store/auth/auth.types';
import loadingSelectors from 'store/api/loading/loading.selectors';
import dialogActions from 'store/ui/dialog/dialog.actions';

type Props = {
  isAuth: boolean,
  isLogoutLoading: boolean,
  onLogout: () => void,
  onShowLoginDialog: () => void,
};

const useStyles = makeStyles((theme) => ({
  root: {
    minWidth: 72,
    textTransform: 'none',
    fontSize: theme.typography.h6.fontSize,
    fontWeight: theme.typography.fontWeightRegular,
    '&$disabled': {
      color: theme.palette.grey[300],
    },
  },
  disabled: {},
  logout: {
    color: theme.palette.common.white,
  },
}));

const AuthButton = (props: Props) => {
  const { isAuth, isLogoutLoading, onLogout, onShowLoginDialog } = props;
  const classes = useStyles();

  return (
    <>
      {isAuth ? (
        <Button
          classes={{ root: classes.root, disabled: classes.disabled, text: classes.logout }}
          disabled={isLogoutLoading}
          onClick={onLogout}
        >
          Logout
        </Button>
      ) : (
        <Button className={classes.root} color="primary" onClick={onShowLoginDialog}>
          Login
        </Button>
      )}
    </>
  );
};

const mapStateToProps = createStructuredSelector({
  isAuth: authSelectors.selectIsAuth,
  isLogoutLoading: loadingSelectors.createIsLoadingSelector([AuthActionTypes.LOGOUT]),
});

const mapDispatchToProps = {
  onLogout: authActions.logout,
  onShowLoginDialog: () => dialogActions.showDialog(DialogTypes.LOGIN),
};

export default connect(mapStateToProps, mapDispatchToProps)(AuthButton);
