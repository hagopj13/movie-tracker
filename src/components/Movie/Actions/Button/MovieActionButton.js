// @flow
import React from 'react';
import type { Node } from 'react';
import { makeStyles } from '@material-ui/core/styles';
import Tooltip from '@material-ui/core/Tooltip';
import Fab from '@material-ui/core/Fab';
import { lightBlue } from '@material-ui/core/colors';

type Props = {
  children: Node,
  title: string,
  onClick: () => void,
};

const useStyles = makeStyles((theme) => ({
  tooltip: {
    fontSize: theme.typography.subtitle2.fontSize,
  },
  fab: {
    backgroundColor: lightBlue[800],
    '&:hover': {
      backgroundColor: lightBlue[900],
    },
  },
}));

const MovieActionButton = (props: Props, ref) => {
  const { children, title, onClick } = props;

  const classes = useStyles();

  return (
    <Tooltip classes={{ tooltip: classes.tooltip }} title={title}>
      <Fab className={classes.fab} ref={ref} size="medium" onClick={onClick}>
        {children}
      </Fab>
    </Tooltip>
  );
};

export default React.forwardRef<Props, Fab>(MovieActionButton);
