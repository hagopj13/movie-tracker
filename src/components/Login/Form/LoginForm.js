// @flow
import React from 'react';
import { useForm } from 'react-hook-form';
import * as yup from 'yup';
import { makeStyles } from '@material-ui/core/styles';
import TextField from '@material-ui/core/TextField';
import Button from '@material-ui/core/Button';
import Typography from '@material-ui/core/Typography';

import PasswordTextField from 'components/Input/PasswordTextField/PasswordTextField';

type Props = {
  isLoginLoading: boolean,
  loginError: string,
  onSubmit: ({ username: string, password: string }) => void,
};

const useStyles = makeStyles((theme) => ({
  root: {
    '& > *': {
      display: 'block',
      marginBottom: theme.spacing(0.5),
    },
  },
  bottomContainer: {
    display: 'flex',
    alignItems: 'center',
    justifyContent: 'space-between',
  },
  errorText: {
    marginLeft: theme.spacing(1.5),
  },
}));

const schema = yup.object().shape({
  username: yup.string().required(),
  password: yup.string().required(),
});

const LoginForm = (props: Props) => {
  const { isLoginLoading, loginError, onSubmit } = props;

  const classes = useStyles();
  const { register, handleSubmit, errors } = useForm({
    mode: 'onSubmit',
    validationSchema: schema,
  });

  return (
    <form onSubmit={handleSubmit(onSubmit)} className={classes.root}>
      <TextField
        name="username"
        inputRef={register}
        autoFocus
        fullWidth
        size="small"
        variant="outlined"
        label="Username"
        error={!!errors.username}
        helperText={errors.username?.message ?? ' '}
      />
      <PasswordTextField
        name="password"
        inputRef={register}
        fullWidth
        size="small"
        variant="outlined"
        label="Password"
        error={!!errors.password}
        helperText={errors.password?.message ?? ' '}
      />
      <div className={classes.bottomContainer}>
        <Typography color="error" className={classes.errorText}>
          {loginError}
        </Typography>
        <Button
          type="submit"
          className={classes.submitButton}
          size="large"
          variant="outlined"
          color="primary"
          disabled={isLoginLoading}
        >
          Login
        </Button>
      </div>
    </form>
  );
};

export default LoginForm;
