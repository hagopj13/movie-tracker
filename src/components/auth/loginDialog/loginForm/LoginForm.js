import React from 'react';
import { useForm } from 'react-hook-form';
import * as yup from 'yup';
import { makeStyles } from '@material-ui/core/styles';
import TextField from '@material-ui/core/TextField';
import Button from '@material-ui/core/Button';

import PasswordTextField from '../../../common/input/passwordTextField/PasswordTextField';

type Props = {
  onSubmit: (data) => void,
};

const useStyles = makeStyles((theme) => ({
  root: {
    '& > *': {
      display: 'block',
      marginBottom: theme.spacing(0.5),
    },
  },
  submitButton: {
    marginLeft: 'auto',
  },
}));

const schema = yup.object().shape({
  username: yup.string().required(),
  password: yup.string().required(),
});

const LoginForm = (props: Props) => {
  const { onSubmit } = props;
  const classes = useStyles();
  const { register, handleSubmit, errors } = useForm({
    mode: 'onSubmit',
    validationSchema: schema,
  });

  return (
    <form onSubmit={handleSubmit(onSubmit)} className={classes.root}>
      <TextField
        name="username"
        inputRef={register}
        autoFocus
        fullWidth
        size="small"
        variant="outlined"
        label="Username"
        error={!!errors.username}
        helperText={errors.username?.message ?? ' '}
      />
      <PasswordTextField
        name="password"
        inputRef={register}
        fullWidth
        size="small"
        variant="outlined"
        label="Password"
        error={!!errors.password}
        helperText={errors.password?.message ?? ' '}
      />
      <Button
        type="submit"
        className={classes.submitButton}
        size="large"
        variant="outlined"
        color="primary"
      >
        Login
      </Button>
    </form>
  );
};

export default LoginForm;
