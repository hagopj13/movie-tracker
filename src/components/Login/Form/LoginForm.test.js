import React from 'react';
import { act } from 'react-dom/test-utils';
import { createShallow, createMount } from '@material-ui/core/test-utils';
import Typography from '@material-ui/core/Typography';
import Button from '@material-ui/core/Button';

import LoginForm from './LoginForm';

describe('LoginForm component', () => {
  let shallow;
  let mount;

  beforeAll(() => {
    shallow = createShallow();
    mount = createMount();
  });

  it('should render the LoginForm component correctly', () => {
    const mockProps = {
      isLoginLoading: false,
      loginError: '',
      onSubmit: jest.fn(),
    };
    const wrapper = shallow(<LoginForm {...mockProps} />);
    expect(wrapper).toMatchSnapshot();
  });

  it('should not disable the login button when login is not loading', () => {
    const mockProps = {
      isLoginLoading: false,
      loginError: '',
      onSubmit: jest.fn(),
    };
    const wrapper = shallow(<LoginForm {...mockProps} />);
    expect(wrapper.find(Button).prop('disabled')).toBe(false);
  });

  it('should disable the login button when login is loading', () => {
    const mockProps = {
      isLoginLoading: true,
      loginError: '',
      onSubmit: jest.fn(),
    };
    const wrapper = shallow(<LoginForm {...mockProps} />);
    expect(wrapper.find(Button).prop('disabled')).toBe(true);
  });

  it('should render the error text when loginError is set', () => {
    const loginError = 'some error';
    const mockProps = {
      isLoginLoading: false,
      loginError,
      onSubmit: jest.fn(),
    };
    const wrapper = shallow(<LoginForm {...mockProps} />);
    expect(wrapper.find(Typography).text()).toBe(loginError);
  });

  it('should call the onSubmit when form is submitted', async () => {
    const mockSubmit = jest.fn();
    const mockProps = {
      isLoginLoading: false,
      loginError: '',
      onSubmit: mockSubmit,
    };
    const wrapper = mount(<LoginForm {...mockProps} />);

    const username = 'someUsername';
    const password = 'somePassword';

    const usernameElement = wrapper.find('input').at(0);
    usernameElement.getDOMNode().value = username;
    usernameElement.getDOMNode().dispatchEvent(new Event('input'));

    const passwordElement = wrapper.find('input').at(1);
    passwordElement.getDOMNode().value = password;
    passwordElement.getDOMNode().dispatchEvent(new Event('input'));

    await act(async () => {
      wrapper.find('form').simulate('submit');
    });

    expect(mockSubmit).toHaveBeenCalledTimes(1);
    expect(mockSubmit.mock.calls[0][0]).toEqual({ username, password });
  });
});
