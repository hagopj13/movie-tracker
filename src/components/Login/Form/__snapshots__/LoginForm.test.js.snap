// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`LoginForm component should render the LoginForm component correctly 1`] = `
<form
  className="makeStyles-root-1"
  onSubmit={[Function]}
>
  <WithStyles(ForwardRef(TextField))
    autoFocus={true}
    error={false}
    fullWidth={true}
    helperText=" "
    inputRef={[Function]}
    label="Username"
    name="username"
    size="small"
    variant="outlined"
  />
  <PasswordTextField
    error={false}
    fullWidth={true}
    helperText=" "
    inputRef={[Function]}
    label="Password"
    name="password"
    size="small"
    variant="outlined"
  />
  <div
    className="makeStyles-bottomContainer-2"
  >
    <WithStyles(ForwardRef(Typography))
      className="makeStyles-errorText-3"
      color="error"
    />
    <WithStyles(ForwardRef(Button))
      color="primary"
      disabled={false}
      size="large"
      type="submit"
      variant="outlined"
    >
      Login
    </WithStyles(ForwardRef(Button))>
  </div>
</form>
`;
